deps/gamma.jl

```diff
@@ -0,0 +1,242 @@
+    */
+    nz = q - p;
+    if (nz != half) {
+      if (nz > half) {
+        p += one;
+        nz = q - p;
+      }
+      nz = m_pi / $tan_str(m_pi * nz);
+    } else {
+      nz = zero;
+    }
+    xi = one - xi;
+  }
+
+  /* use the recurrence psi(x+1) = psi(x) + 1/x. */
+  s = xi;
+  w = zero;
+  while (s < 10.0) {
+    w += one / s;
+    s += one;
+  }
+
+  yi = digamma_impl_maybe_poly_$F(s);
+
+  yi = $log_str(s) - (half / s) - yi - w;
+
+  return (negative) ? yi - nz : yi;
+
+}
+
+__device__ __host__ int zeta_impl_series_$F($T *a, $T *b, $T *s, const $T x,
+                                         const $T machep) {
+  int i = 0;
+  while ((i < 9) $zeta_impl_series_2nd_cond) {
+    i += 1;
+    *a += $one_str;
+    *b = $pow_str(*a, -x);
+    *s += *b;
+    if ($fabs_str(*b / *s) < machep) {
+      return true;
+    }
+  }
+
+  // Return whether we are done
+  return false;
+}
+
+__device__ __host__ $T zeta_impl_$F($T x, $T q) {
+  int i;
+  $T p, r, a, b, k, s, t, w;
+
+  const $T A[] = {
+      12.0,
+      -720.0,
+      30240.0,
+      -1209600.0,
+      47900160.0,
+      -1.8924375803183791606e9, /*1.307674368e12/691*/
+      7.47242496e10,
+      -2.950130727918164224e12,  /*1.067062284288e16/3617*/
+      1.1646782814350067249e14,  /*5.109094217170944e18/43867*/
+      -4.5979787224074726105e15, /*8.028576626982912e20/174611*/
+      1.8152105401943546773e17,  /*1.5511210043330985984e23/854513*/
+      -7.1661652561756670113e18  /*1.6938241367317436694528e27/236364091*/
+  };
+
+  const $T maxnum = $max_str;
+  const $T zero = 0.0, half = 0.5, one = 1.0;
+  const $T machep = 1e-15;
+
+  if (x == one) return maxnum;
+
+  if (x < one) {
+    return zero;
+  }
+
+  if (q <= zero) {
+    if (q == $floor_str(q)) {
+      return maxnum;
+    }
+    p = x;
+    r = $floor_str(p);
+    if (p != r) return zero;
+  }
+
+  /* Permit negative q but continue sum until n+q > +9 .
+   * This case should be handled by a reflection formula.
+   * If q<0 and x is an integer, there is a relation to
+   * the polygamma function.
+   */
+  s = $pow_str(q, -x);
+  a = q;
+  b = zero;
+
+  // Run the summation in a helper function that is specific to the floating
+  // precision
+  if (zeta_impl_series_$F(&a, &b, &s, x, machep)) {
+    return s;
+  }
+
+  w = a;
+  s += b * w / (x - one);
+  s -= half * b;
+  a = one;
+  k = zero;
+  for (i = 0; i < 12; i++) {
+    a *= x + k;
+    b /= w;
+    t = a * b / A[i];
+    s = s + t;
+    t = $fabs_str(t / s);
+    if (t < machep) return s;
+    k += one;
+    a *= x + k;
+    b /= w;
+    k += one;
+  }
+  return s;
+};
+
+__device__ __host__ $T gamma_impl_$F($T x) {
+  return tgamma(x);
+}
+
+__device__ __host__ $T polygamma_impl_$F(int n, $T x) {
+  if (n == 0) {
+    return digamma_impl_$F(x);
+  }
+
+  // dumb code to calculate factorials
+  $T factorial = $one_str;
+  for (int i = 0; i < n; i++) {
+    factorial *= (i + 1);
+  }
+
+  $T s = n % 2 == 0 ? -$one_str : $one_str;
```

Comment:
Just for people who wonder why we had numerical errors before: this line was coded as `s = powf(-1.0, n + 1)` in the original code for float. This causes numerical errors somehow. I changed it to avoid using unnecessary power function to avoid this.